servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcUtils.java

```diff
@@ -138,50 +150,56 @@ static void setStatus(final HttpHeaders trailers, final Throwable cause, final B
 
     static <Resp> Publisher<Resp> validateResponseAndGetPayload(final StreamingHttpResponse response,
                                                                 final HttpDeserializer<Resp> deserializer) {
-        // In case of server error, gRPC may return only one HEADER frame with endStream=true. Our
-        // HTTP1-based implementation translates them into response headers so we need to look for
-        // the status in both headers and trailers. Since this is streaming response and we have the headers now, we
-        // check for error here first. If we see trailers later in payloadBodyAndTrailers(), we will check for error
-        // there.
-        final HttpHeaders respHeaders = response.headers();
-        GrpcStatusException grpcStatusException = extractGrpcExceptionFromHeaders(respHeaders);
-        if (grpcStatusException != null) {
-            return Publisher.failed(grpcStatusException);
+        // In case of an empty response, gRPC-server may return only one HEADER frame with endStream=true. Our
+        // HTTP1-based implementation translates them into response headers so we need to look for a grpc-status in both
+        // headers and trailers. Since this is streaming response and we have the headers now, we check for the
+        // grpc-status here first. If there is no grpc-status in headers, we look for it in trailers later.
+        final HttpHeaders headers = response.headers();
+        final GrpcStatusCode grpcStatusCode = extractGrpcStatusCodeFromHeaders(headers);
+        if (grpcStatusCode != null) {
+            final GrpcStatusException grpcStatusException = convertToGrpcStatusException(grpcStatusCode, headers);
+            return response.payloadBodyAndTrailers().ignoreElements()
+                    .concat(grpcStatusException != null ? failed(grpcStatusException) : empty());
         }
-        return deserializer.deserialize(respHeaders, response.payloadBodyAndTrailers().map(o -> {
-            if (o instanceof HttpHeaders) {
-                // We have already checked for error in headers above, now we just check in trailers.
-                GrpcStatusException ex = extractGrpcExceptionFromHeaders((HttpHeaders) o);
-                if (ex != null) {
-                    throw ex;
-                }
-            } else if (!(o instanceof Buffer)) {
-                throw new IllegalArgumentException("Unexpected payload type: " + o.getClass());
-            }
-            return o;
-        }).filter(o -> !(o instanceof HttpHeaders)).map(o -> (Buffer) o));
+
+        response.transformRaw(ENSURE_GRPC_STATUS_RECEIVED);
+        return deserializer.deserialize(headers, response.payloadBodyAndTrailers()
```

Comment:
This helps to catch the case when server does not send trailers at all. 
In this case `payloadBodyAndTrailers()` returns only `Buffer`s. `transformRaw` helps to create empty trailers and catch that there is no `grpc-status`.
I did not add a comment here, because the test will fail if this changes, but I can add it if it will be useful.

Reply:
Good idea, I will keep it with `transformRaw` for now and we can optimize if it show ups as a problem later.